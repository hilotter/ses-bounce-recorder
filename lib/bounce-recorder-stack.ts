import cdk = require('@aws-cdk/core');
import sns = require('@aws-cdk/aws-sns');
import subs = require('@aws-cdk/aws-sns-subscriptions');
import dynamodb = require('@aws-cdk/aws-dynamodb')
import lambda = require('@aws-cdk/aws-lambda')
import apigw = require('@aws-cdk/aws-apigateway');
import { ApiKeySourceType } from '@aws-cdk/aws-apigateway';
import { read } from 'fs';

export class BounceRecorderStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const topicName = process.env.BOUNCE_RECORDER_SNS_TOPIC || 'ses-notifications';
    const topic = new sns.Topic(this, topicName, {
      displayName: 'SES bounce notifications'
    })

    const tableName = process.env.BOUNCE_RECORDER_DYNAMODB_TABLE || 'ses-mailing'
    const table = new dynamodb.Table(this, tableName, {
      partitionKey: { name: 'UserId', type: dynamodb.AttributeType.STRING }
    });

    const recordHandler = new lambda.Function(this, 'BounceRecorderHandler', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'bounce_recorder.handler',
      code: lambda.Code.asset('lambda'),
      timeout: cdk.Duration.seconds(30),
      environment: {
        DYNAMODB_TABLE_NAME: table.tableName
      }
    });

    table.grantReadWriteData(recordHandler);
    topic.addSubscription(new subs.LambdaSubscription(recordHandler));

    const scanLimit = process.env.BOUNCE_RECORDER_SCAN_LIMIT || '1000';
    const readHandler = new lambda.Function(this, 'BounceReaderHandler', {
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: 'bounce_reader.handler',
      code: lambda.Code.asset('lambda'),
      timeout: cdk.Duration.seconds(300),
      environment: {
        DYNAMODB_TABLE_NAME: table.tableName,
        DYNAMODB_SCAN_LIMIT: scanLimit
      }
    });

    table.grantReadData(readHandler);

    const api = new apigw.RestApi(this, 'BounceReaderEndpoint', {});
    const integration = new apigw.LambdaIntegration(readHandler);
    const echo = api.root.addMethod('GET', integration, {
      apiKeyRequired: true
    })
    const apiKey = api.addApiKey('BounceReaderApiKey');
    const plan = api.addUsagePlan('UsagePlan', {
      name: 'BounceReader',
      apiKey: apiKey
    });

    plan.addApiStage({
      stage: api.deploymentStage,
      throttle: []
    });
  }
}
