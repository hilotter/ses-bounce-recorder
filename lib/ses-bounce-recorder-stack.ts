import * as cdk from '@aws-cdk/core';
import * as sns from '@aws-cdk/aws-sns';
import * as subs from '@aws-cdk/aws-sns-subscriptions';
import * as lambda from '@aws-cdk/aws-lambda'
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs'

export class SesBounceRecorderStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    // NOTE: Please set SNS topic subscribe in SES notification manually
    // - Bounce Notifications SNS Topic
    // - Complaint Notifications SNS Topic
    const topicName = process.env.BOUNCE_RECORDER_SNS_TOPIC || 'ses-notifications'
    const topic = new sns.Topic(this, topicName, {
      topicName: topicName,
      displayName: 'SES bounce notifications'
    })

    const bounceRecorderLambda = new NodejsFunction(
      this,
      'BounceRecorderLambda',
      {
        functionName: `SesBounceRecorder`,
        entry: 'src/lambda/handlers/sesBounceRecorder.ts',
        runtime: lambda.Runtime.NODEJS_12_X,
        timeout: cdk.Duration.seconds(30),
      }
    )
    bounceRecorderLambda.addEnvironment('WEBHOOK_KEY', process.env.WEBHOOK_KEY || '')
    bounceRecorderLambda.addEnvironment('WEBHOOK_URL', process.env.WEBHOOK_URL || '')

    topic.addSubscription(new subs.LambdaSubscription(bounceRecorderLambda))
  }
}
