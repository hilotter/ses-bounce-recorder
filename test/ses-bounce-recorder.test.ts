import { Template } from "aws-cdk-lib/assertions"
import * as cdk from "aws-cdk-lib"
import { SesBounceRecorderStack } from "../lib/ses-bounce-recorder-stack"

describe("SesBounceRecorderStack", () => {
  const OLD_ENV = process.env

  beforeEach(() => {
    jest.resetModules()
    process.env = { ...OLD_ENV }
    process.env.BOUNCE_RECORDER_SNS_TOPIC = "ses-notifications"
    process.env.WEBHOOK_KEY = "test-key"
    process.env.WEBHOOK_URL = "https://example.com"
  })

  afterAll(() => {
    process.env = OLD_ENV
  })

  test("sns", () => {
    const template = getTemplate()
    template.hasResourceProperties("AWS::SNS::Topic", {
      TopicName: "ses-notifications",
    })
  })

  test("lambda", () => {
    const template = getTemplate()
    template.hasResourceProperties("AWS::Lambda::Function", {
      FunctionName: "SesBounceRecorder",
      Runtime: "nodejs16.x",
      Environment: {
        Variables: {
          WEBHOOK_KEY: "test-key",
          WEBHOOK_URL: "https://example.com"
        }
      },
    })
  })

  test("snapshot", () => {
    const template = getTemplate().toJSON()
    expect(template).toMatchSnapshot()
  })
})

function getTemplate(): Template {
  const app = new cdk.App()

  const sesBounceRecorderStack = new SesBounceRecorderStack(app, "SesBounceRecorderStack")

  return Template.fromStack(sesBounceRecorderStack)
}
