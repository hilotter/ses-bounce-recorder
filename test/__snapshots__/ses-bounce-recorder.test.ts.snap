// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SesBounceRecorderStack snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BounceRecorderLambda17BFDD3A": Object {
      "DependsOn": Array [
        "BounceRecorderLambdaServiceRoleB5D17182",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "STUB-HASH.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "WEBHOOK_KEY": "test-key",
            "WEBHOOK_URL": "https://example.com",
          },
        },
        "FunctionName": "SesBounceRecorder",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "BounceRecorderLambdaServiceRoleB5D17182",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BounceRecorderLambdaAllowInvokeSesBounceRecorderStacksesnotifications19C3D3BFD558C6DC": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "BounceRecorderLambda17BFDD3A",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "sesnotifications21CEB3A8",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "BounceRecorderLambdaServiceRoleB5D17182": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BounceRecorderLambdasesnotifications9F9F6621": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "BounceRecorderLambda17BFDD3A",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "sesnotifications21CEB3A8",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "sesnotifications21CEB3A8": Object {
      "Properties": Object {
        "DisplayName": "SES bounce notifications",
        "TopicName": "ses-notifications",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
